File: src/main/java/edu/cornell/cs/apl/viaduct/Variable.java
package edu.cornell.cs.apl.viaduct;

/** represents referenced or declared variables. */
public class Variable {
  String name;

  public Variable(String name) {
    this.name = name;
  }

  public String getName() {
    return this.name;
  }

  public String toString() {
    return this.name;
  }

  @Override
  public boolean equals(Object o) {
    if (o instanceof Variable) {
      Variable vo = (Variable) o;
      return this.name.equals(vo.getName());

    } else {
      return false;
    }
  }

  @Override
  public int hashCode() {
    return this.name.hashCode();
  }
}

File: src/main/kotlin/edu/cornell/cs/apl/viaduct/syntax/Variable.kt
package edu.cornell.cs.apl.viaduct.syntax

/** A variable is a name that stands for a value or an object instance. */
sealed class Variable : Name

/**
 * A variable that binds base values.
 *
 * Temporaries are generated internally to name expression results.
 */
data class Temporary(override val name: String) : Variable() {
    override val nameCategory: String
        get() = "temporary"
}

/** A variable that binds an object. */
data class ObjectVariable(override val name: String) : Variable() {
    override val nameCategory: String
        get() = "object"
}
